### ISSUE TITLE
TypeError: Cannot read property 'dispose' of undefined

### ISSUE THREAD

**Issue by JeremyGrieshop** [SurenAt93/monaco-react#6]
Hi, I've been playing around with your React integration of Monaco.  Thanks for all your hard work in doing this!

I ran into a few problems when the editor unmounts.  I'm creating the editor within a functional component and sometimes when I click something that would cause the editor to unmount (I presume), I get errors such as:

```
  TypeError: Cannot read property 'updateOptions' of undefined
(anonymous function)
src/afw_app/admin/node_modules/@monaco-editor/react/lib/Editor/Editor.js:166
  163 |   monacoRef.current.editor.setTheme(theme);
  164 | }, [theme]);
  165 | (0, _hooks.useUpdate)(function (_) {
> 166 |   editorRef.current.updateOptions(options);
      | ^  167 | }, [options]);
  168 | 
  169 | function createEditor() {
```

And:

```
TypeError: Cannot read property 'dispose' of undefined
removeEditor
src/afw_app/admin/node_modules/@monaco-editor/react/lib/Editor/Editor.js:182
  179 | }
  180 | 
  181 | function removeEditor() {
> 182 |   editorRef.current.dispose();
      | ^  183 | }
  184 | 
  185 | return _react["default"].createElement("section", {
```

I have a few other questions/comments, that I'll leave in other open issues.

**Comment by JeremyGrieshop** [comment]
FWIW, I know this isn't likely a permanent solution, but just adding an "if editorRef.current" and "if monacoRef.current" fixed this temporarily for me.

**Comment by SurenAt93** [comment]
Thank you for taking the time to open this issue. 👍

I’ll try to look at it tomorrow (also the other issue you opened). I’ll keep you up to date.
Before, I’ll kindly ask you to create a minimal [codesandbox](https://codesandbox.io) reproducing the issue. Thank you!

**Comment by JeremyGrieshop** [comment]
The project I'm using is fairly complex, but I intend to spend some time getting the problem I'm seeing isolated to a sandbox that you can look at.

**Comment by JeremyGrieshop** [comment]
Ok, this is probably a very bizarre example, but it seems to do enough to reliably reproduce the problem.  You can open the dev console on your browser to see the updateOptions error after ~5 seconds.  I setup a hook to update the state every 5 seconds to try and break it.  

https://codesandbox.io/embed/eager-hellman-bui7v

From my own debugging, it seems that because there are two editor components, there are two CreateEditor() calls.  When one of them gets the removeEditor() call, the second call to removeEditor() will throw the exception, presumably because the first one already removed the ref.  

If you don't have much time to work on it, that's understandable.  I could probably fork/work on it myself, but just didn't want to duplicate efforts.  Thanks again.

**Comment by SurenAt93** [comment]
I will look at it more detailed. There is definitely a bug; at least it should show a more descriptive error message.

**Comment by onatm** [comment]
This also happens when refreshing a page that has an `Editor` in it. I'm debugging it atm to find out the root cause.

**Comment by SurenAt93** [comment]
Right now the biggest problem is descriptive error messages; I’ll try to do it first.

**Comment by SurenAt93** [comment]
Okay. The issue was revealed. The problem is: you are interacting with the editor before its initialization. Let me explain what happens in your provided example:

The first editor mounts, it calls “setSmth” which leads re-render, (at this particular time the second one is not mounted yet) during re-render the second editor receives a new options prop (very interesting part here: as you passed the value of `options` prop `{}`, every time it’s a new object); and as the `options` props changed the second editor tries to update editor options, which is not mounted yet.

**Comment by SurenAt93** [comment]
There was an issue, where we have already talked about this topic. [Here it is](https://github.com/SurenAt93/monaco-react/issues/1).

**Comment by SurenAt93** [comment]
Possible solutions: there was an option to check if the editor is mounted or not inside of the editor to avoid such kind of things. But we avoided do it, and plus it's not just "check". I thought to add an error message about interacting before editor mount, but now I am thinking about checking.

One more note: checks you are doing [here](https://github.com/JeremyGrieshop/react-monaco-editor/blob/master/src/lib/Editor.js#L39) will lead to inconsistency. Let me explain: someone passes, for example, "something"  to `value` prop when the editor is not mounted yet. You are checking the existence of the editor before setting the new value. Okay, the "if" condition is falsy and value is not set. A little bit later Editor will be mounted, but not with "something" inside.

**Comment by JeremyGrieshop** [comment]
> Possible solutions: there was an option to check if the editor is mounted or not inside of the editor to avoid such kind of things. But we avoided do it, and plus it's not just "check". I thought to add an error message about interacting before editor mount, but now I am thinking about checking.
> 
> One more note: checks you are doing [here](https://github.com/JeremyGrieshop/react-monaco-editor/blob/master/src/lib/Editor.js#L39) will lead to inconsistency. Let me explain: someone passes, for example, "something" to `value` prop when the editor is not mounted yet. You are checking the existence of the editor before setting the new value. Okay, the "if" condition is falsy and value is not set. A little bit later Editor will be mounted, but not with "something" inside.

I kinda see what you're saying, but, it seems that later when the Editor does get mounted, it passes "value" to the constructor:  https://github.com/JeremyGrieshop/react-monaco-editor/blob/master/src/lib/Editor.js#L20

The logic is a bit confusing for the massive amount of useEffect() calls that are there, and it probably warrants a useReducer to get the state machine right.

**Comment by SurenAt93** [comment]
Okay, let me examine your solution more detailed. Also, I'll share the new version with the solution here; let it be open.

**Comment by SurenAt93** [comment]
The issue has already been fixed, I’ll try to publish it today. Will keep you up to date here.

**Comment by SurenAt93** [comment]
The new version has already been published (**1.0.5**). @JeremyGrieshop @onatm I'll kindly ask you to check the new version in your cases. And thank you for the help.

**Comment by onatm** [comment]
It works like a charm @SurenAt93. Thank you for the fix.

My initial problem was related to some unnecessary mounting and unmounting related to `react-router`. However, the same issue happened when I added some `options` to the editor after I fixed the routing issues. Now all the issues are fixed 🎉

**Comment by SurenAt93** [comment]
> It works like a charm @SurenAt93. Thank you for the fix.
> 
> My initial problem was related to some unnecessary mounting and unmounting related to `react-router`. However, the same issue happened when I added some `options` to the editor after I fixed the routing issues. Now all the issues are fixed 🎉

Glad to hear that :ok_hand:

**Comment by JeremyGrieshop** [comment]
> The new version has already been published (**1.0.5**). @JeremyGrieshop @onatm I'll kindly ask you to check the new version in your cases. And thank you for the help.

The CodeSandbox example seems to work with 1.0.5, thanks!

**Comment by SurenAt93** [comment]
It's also done for DiffEditor.
